text/3301-atomic-memcpy.md

```diff
@@ -0,0 +1,227 @@
+- Feature Name: `atomic_memcpy`
+- Start Date: 2022-08-14
+- RFC PR: [rust-lang/rfcs#3301](https://github.com/rust-lang/rfcs/pull/3301)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+This is a proposal to add `AtomicPerByte<T>`, to represent _tearable atomics_.
+This makes it possible to properly implement a _sequence lock_ in Rust.
+
+# The Problem
+
+It's currently not possible to implement an efficient and perfectly
+(theoretically) correct sequence lock in Rust.
+
+Unlike most locking mechanisms, a sequence lock doesn't prevent a race
+to access the data it projects.
+Instead, it detects a race only after the load operation already happened,
+and retries it if the load operation raced with a write operation.
+
+A sequence lock in Rust looks something like this:
+
+```rust
+// Incomplete example
+
+pub struct SeqLock<T> {
+    seq: AtomicUsize,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl Sync<T: Copy + Send> for SeqLock<T> {}
+
+impl<T: Copy> SeqLock<T> {
+    /// Safety: Only call from one thread.
+    pub unsafe fn write(&self, value: T) {
+        self.seq.fetch_add(1, Relaxed);
+        write_data(&mut self.data, value, Release);
+        self.seq.fetch_add(1, Release);
+    }
+
+    pub fn read(&self) -> T {
+        loop {
+            let s1 = self.seq.load(Acquire);
+            let data = read_data(&self.data, Acquire);
+            let s2 = self.seq.load(Relaxed);
```

Comment:
There's something very subtle here that I had not appreciated until a few weeks ago: we have to ensure that the `load` here cannot return an outdated value that would prevent us from noticing a seqnum bump.

The reason this is the case is that if there is a concurrent `write`, and if and part of `data` reads from that write, then we have a release-acquire pair, so then we are guaranteed to see at least the first `fetch_add` from `write`, and thus we will definitely see a version conflict. OTOH if the `s1` reads-from some second `fetch_add` in `write`, then *that* forms a release-acquire pair, and we will definitely see the data.

So, all the release/acquire are necessary here. (I know this is not a seqlock tutorial, and @m-ou-se is certainly aware of this, but it still seemed worth pointing out -- many people reading this will not be aware of this.)

Reply:
Yeah exactly. This is why people are sometimes asking for a "release-load" operation. This second load operation needs to happen "after" the `read_data()` part, but the usual (incorrect) `read_data` implementation doesn't involve atomic operations or a memory ordering, so they attempt to solve this issue with a memory ordering on that final load, which isn't possible. The right solution is a memory ordering on the `read_data()` operation.

Reply:
Under a reordering based atomic model (as CPUs use), a release load makes sense and works. Release loads don't _really_ work unless they are also RMWs (`fetch_add(0)`) under the C11 model.

Reply:
Yeah, the famous [seqlock paper](https://www.hpl.hp.com/techreports/2012/HPL-2012-68.pdf) discusses "read dont-modify write" operations.